File: Python/0121-e-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0
        best = float('-inf')
        minPrice = prices[0]
        for i in range(len(prices)):
            minPrice = min(minPrice, prices[i])
            best = max(best, prices[i] - minPrice)
        return best

File: leetcode/C++/0121-e-best-time-to-buy-and-sell-stock.cpp
// Prefix min
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if (n == 0) return 0;
        int res = 0;
        int prefix_min = prices[0];
        for (int i = 1; i < n; ++i) {
            res = max(res, prices[i] - prefix_min);
            prefix_min = min(prefix_min, prices[i]);
        }
        return res;
    }
};